core/src/main/scala/kafka/log/Log.scala

```diff
@@ -465,6 +473,9 @@ class Log(@volatile var dir: File,
       // check that offsets are monotonically increasing
       if(lastOffset >= entry.offset)
         monotonic = false
+      // keep track of the highest offset seen
+      if (maxOffset < entry.offset)
+        maxOffset = entry.offset
```

Comment:
Got it. I thought because of the check on line 474/475 that it was accepted for offsets to come in out of order, but I see on line 405 if we are taking the offsets we are given, we check that the offsets are monotonic, and throw an exception otherwise. Ill remove this new field and book-keeping and just use lastOffset.